src/librustc_mir/monomorphize/polymorphize.rs

```diff
@@ -0,0 +1,287 @@
+
+/// Provide implementations of queries relating to polymorphization analysis.
+pub fn provide(providers: &mut Providers<'_>) {
+    providers.used_generic_params = used_generic_params;
+}
+
+/// Determine which generic parameters are used by the function/method/closure represented by
+/// `def_id`. Returns a `BitSet` with a domain size matching the number of generic parameters,
+/// set bits represent that the matching generic parameter is used.
+fn used_generic_params(tcx: TyCtxt<'_>, def_id: DefId) -> BitSet<u32> {
+    debug!("used_generic_params({:?})", def_id);
+    let generics = tcx.generics_of(def_id);
+    debug!("used_generic_params: generics={:?}", generics);
+    used_generic_params_inner(tcx, def_id, generics)
+        .unwrap_or_else(|| BitSet::new_filled(generics.count()))
+}
+
+fn used_generic_params_inner<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    generics: &'tcx ty::Generics,
+) -> Option<BitSet<u32>> {
+    // Exit early when there are no parameters to be unused.
+    if generics.count() == 0 {
+        return None;
+    }
+
+    // Exit early when there is no MIR available.
+    if !tcx.is_mir_available(def_id) {
+        debug!("used_generic_params: (no mir available) def_id={:?}", def_id);
+        return None;
+    }
+
+    let mut used_parameters = BitSet::new_empty(generics.count());
+    debug!(
+        "used_generic_params: used_parameters.domain_size()={:?}",
+        used_parameters.domain_size()
+    );
+
+    mark_used_by_default_parameters(tcx, def_id, generics, &mut used_parameters);
+
+    // Visit MIR and accumululate used generic parameters.
+    let body = tcx.optimized_mir(def_id).unwrap_read_only();
+    let mut vis =
+        UsedGenericParametersVisitor { tcx, def_id, used_parameters: &mut used_parameters };
+    vis.visit_body(body);
+    debug!("used_generic_params: used_parameters={:?}", used_parameters);
+
+    mark_used_by_predicates(tcx, def_id, &mut used_parameters);
+
+    // Emit errors for debugging and testing if enabled.
+    let is_full = used_parameters.count() == used_parameters.domain_size();
+    if tcx.sess.opts.debugging_opts.polymorphize_errors && !is_full {
+        emit_used_generic_params_error(tcx, def_id, generics, &used_parameters);
+    }
+
+    Some(used_parameters)
+}
+
+/// Some parameters are considered used-by-default, such as non-generic parameters and the dummy
+/// generic parameters from closures, this function marks them as used. `leaf_is_closure` should
+/// be `true` if the item that `used_generic_params` was invoked on is a closure.
+fn mark_used_by_default_parameters<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    generics: &'tcx ty::Generics,
+    used_parameters: &mut BitSet<u32>,
+) {
+    for param in &generics.params {
+        debug!("mark_used_by_default_parameters: param={:?}", param);
+        // Consider any of the dummy generic parameters that closures have as used.
+        if param.def_id == def_id {
```

Comment:
This could probably be more explicit, although we'd need to make sure `ClosureSubsts` abstracts it at least partially.

One way would be to make these associated `const`s (`SYNTHETIC_PARAM_NAMES`) on `GeneratorSubsts` and `ClosureSubsts`:
https://github.com/rust-lang/rust/blob/55299b2ba99432d10f925cd28ff52fe397577371/src/librustc_typeck/collect.rs#L1376-L1379

That way, calling `.len()` on them is going to always be correct, instead of hardcoding the current values.

Reply:
Fixed in e9eb863.